RFC/RFC-14-p2p-adapter.md

```diff
@@ -0,0 +1,348 @@
+type Route = string[] | 'server'
+type PeerRoutingTable = Map<string, Route>
+
+function calculateRoutingTables(mesh: Map<string, Set<string>>) {
+  const routingTables = new Map<string, PeerRoutingTable>()
+
+  const getOrCreateRoutingTable = (peerId: string) => {
+    let table = routingTables.get(peerId)
+    if (!table) {
+      table = new Map<string, Route>()
+      routingTables.set(peerId, table)
+    }
+    return table
+  }
+
+  function calculateRouteBeetwen(fromPeer: string, toPeer: string, _excluding: string[]): Route {
+    const excluding = new Set<string>(_excluding)
+
+    const calculatedRoutes = getOrCreateRoutingTable(fromPeer)
+
+    const calculatedRoute = calculatedRoutes.get(toPeer)
+    if (calculatedRoute) {
+      return calculatedRoute
+    }
+
+    let route: Route = 'server'
+
+    const fromPeerConnections = mesh.get(fromPeer)
+    if (!fromPeerConnections) {
+      route = 'server'
+    } else if (fromPeerConnections?.has(toPeer)) {
+      route = []
+    } else {
+      for (const p of fromPeerConnections) {
+        if (excluding.has(p)) {
+          continue
+        }
+        let relayedRoute = calculateRouteBeetwen(p, toPeer, [p, ..._excluding])
+        if (relayedRoute !== 'server') {
+          relayedRoute = [p, ...relayedRoute]
+          if (route === 'server' || route.length > relayedRoute.length) {
+            route = relayedRoute
+          }
+        }
+      }
+    }
+
+    calculatedRoutes.set(toPeer, route)
+
+    // NOTE: routes are bidirectional
+    getOrCreateRoutingTable(toPeer).set(fromPeer, route === 'server' ? route : Array.from(route).reverse())
+    return route
+  }
+
+  const peers = new Set<string>()
+
+  for (const [peer, connections] of mesh) {
+    peers.add(peer)
+    for (const connection of connections) {
+      peers.add(connection)
+    }
+  }
+
+  for (const peerFrom of peers) {
+    for (const peerTo of peers) {
+      if (peerFrom === peerTo) {
+        continue
+      }
+
+      calculateRouteBeetwen(peerFrom, peerTo, [])
+    }
+  }
+
+  return routingTables
+}
+```
+
+## Packet
+
+A packet contains a source (peer id) and map specifying to whom the packet is for and the route to follow to reach it.
+
+```typescript
+
+type Packet = {
+  source: Address
+  to: Record<Address, Route>
+}
+```
+
+## Peer
+
+The peer will send a message to every known peer using the routing table, sending it once to each peer. Let's say peer1 is connected to peer2 and peer2 to peer3, when broadcasting to the network peer1 will send:
+
+```typescript
+{ 
+  source: 'peer1', 
+  to: { 'peer2': [], 'peer3': ['peer2'] }
+}
+```
+
+This means each peer will check the `to` field, if they are one of the recipent of the messages the message will be processed, otherwise it will be relayed in the way the route value indicates.
+
+Notice the peer relaying the message is not expected to remove itself either as a recipent or as a hop in the route, since this will require encoding the package again. 
```

Comment:
So, let's suppose the intermediate peer has a connection to the peer from which it received the message... it has to be clever enough to parse and pick up from anything after itself, right?

Reply:
yes, only if in the path to other, so say you have the following mesh:

`a<->b<->c` and `a<->d`

a will send the following message:


```typescript
{
  source: 'a'
  to: {
  'b': [],
  'c': ['b'],
  'd': []
  }
}
```


d will receive it, parse it, and process it
c will receive it, parse it, and process it. But attempt to relay to c as well, because is the route to d. If the connection is cut, it will try to reach d thought the messaging service.

I tried to illustrate this on the peer is disconnected example, but let me know if you think I'm missing something